proof/refine/AARCH64/ArchRetype_R.thy

```diff
@@ -30,109 +32,80 @@ where
+lemmas APIType_capBits_def = APIType_capBits_raw_def[simplified APIType_capBits_gen_def]
+
+lemma APIType_capBits_generic[Retype_R_assms, simp]:
+  "APIType_capBits (APIObjectType api) us = APIType_capBits_gen api us"
+  by (simp add: APIType_capBits_raw_def)
+
+definition makeObjectKO :: "bool \<Rightarrow> domain \<Rightarrow> (kernel_object + AARCH64_H.object_type) \<rightharpoonup> kernel_object"
+  where
+  makeObjectKO_raw_def:
   "makeObjectKO dev d ty \<equiv> case ty of
       Inl KOUserData \<Rightarrow> Some KOUserData
     | Inl (KOArch (KOASIDPool _)) \<Rightarrow> Some (KOArch (KOASIDPool makeObject))
     | Inl (KOArch (KOVCPU _)) \<Rightarrow> Some (KOArch (KOVCPU makeObject))
     | Inr VCPUObject \<Rightarrow> Some (KOArch (KOVCPU makeObject))
-    | Inr (APIObjectType ArchTypes_H.TCBObject) \<Rightarrow> Some (KOTCB (tcbDomain_update (\<lambda>_. d) makeObject))
-    | Inr (APIObjectType ArchTypes_H.EndpointObject) \<Rightarrow> Some (KOEndpoint makeObject)
-    | Inr (APIObjectType ArchTypes_H.NotificationObject) \<Rightarrow> Some (KONotification makeObject)
-    | Inr (APIObjectType ArchTypes_H.CapTableObject) \<Rightarrow> Some (KOCTE makeObject)
+    | Inr (APIObjectType api) \<Rightarrow> makeObjectKO_gen d api
     | Inr PageTableObject \<Rightarrow> Some (KOArch (KOPTE makeObject))
     | Inr VSpaceObject \<Rightarrow> Some (KOArch (KOPTE makeObject))
     | Inr SmallPageObject \<Rightarrow> Some (if dev then KOUserDataDevice else KOUserData)
     | Inr LargePageObject \<Rightarrow> Some(if dev then KOUserDataDevice else KOUserData)
     | Inr HugePageObject \<Rightarrow> Some (if dev then KOUserDataDevice else KOUserData)
     | _ \<Rightarrow> None"
 
-text \<open>makeObject etc. lemmas\<close>
+(* inside of Arch, we don't need to isolate generic component *)
+lemmas makeObjectKO_def = makeObjectKO_raw_def[simplified makeObjectKO_gen_def]
 
-lemma NullCap_valid' [iff]: "s \<turnstile>' capability.NullCap"
-  unfolding valid_cap'_def by simp
+lemma makeObjectKO_generic[Retype_R_assms, simp]:
+  "makeObjectKO dev d (Inr (APIObjectType api)) = makeObjectKO_gen d api"
+  by (simp add: makeObjectKO_raw_def)
 
-lemma valid_obj_makeObject_cte [simp]:
-  "valid_obj' (KOCTE makeObject) s"
-  unfolding valid_obj'_def valid_cte'_def
-  by (clarsimp simp: makeObject_cte)
+text \<open>makeObject etc. lemmas\<close>
 
-lemma valid_obj_makeObject_tcb [simp]:
+lemma valid_obj_makeObject_tcb[Retype_R_assms, simp]:
   "valid_obj' (KOTCB makeObject) s"
   unfolding valid_obj'_def valid_tcb'_def valid_tcb_state'_def valid_arch_tcb'_def
   by (clarsimp simp: makeObject_tcb makeObject_cte tcb_cte_cases_def minBound_word newArchTCB_def
                      cteSizeBits_def)
 
-lemma valid_obj_makeObject_tcb_tcbDomain_update [simp]:
+lemma valid_obj_makeObject_tcb_tcbDomain_update[simp]:
   "d \<le> maxDomain \<Longrightarrow> valid_obj' (KOTCB (tcbDomain_update (\<lambda>_. d) makeObject)) s"
```

Comment:
Similarly to `valid_obj_makeObject_tcb`, not sure if it's worth doing but this is another lemma that feels like it could be made generic with a lower-level interface.

Reply:
well spotted, this can be made generic it seems. done on AARCH64 so far, will have to update the rest